### PR TITLE
[headerssetter] Extend  headerssetter with header actions.

### LINKED ISSUE #7596: Ability to strip headers for AWS Proxy
**Is your feature request related to a problem? Please describe.**
If we deploy a istio (envoy) proxy in front of the collector,  the signing process fails because proxy sets x-forwarded-proto to 'http' causing a mismatch in the signing process.  

Similar to aws-sig4-proxy,  we need the ability to strip out or exclude headers from signing.  See https://github.com/awslabs/aws-sigv4-proxy#examples

**Describe the solution you'd like**
Ability to pass an environment variable/configuration to exclude headers that would include headers that needs to be removed when using awsproxy.  The default value of this configuration can include headers like x-forwarded-port,x-forwarded-proto

**Describe alternatives you've considered**
We are considering suppressing this header using envoy filters.

**Additional context**
To reproduce this issue,  you can simply exec inside otel-collector,  and run the curl command 
```
curl -X POST \
-H 'x-forwarded-proto: http' \
http://localhost:2000/GetSamplingRules
```

### LINKED ISSUE #16581: [extension/headerssetter] Extend the headers setter extension with header modification actions.
### Component(s)

extension/headerssetter

### Is your feature request related to a problem? Please describe.

There will be cases when users would have to insert, update or drop certain headers, see the scenario
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7596#issuecomment-1326484215 where users have to drop a specific header. The headers setter extension could be a good place where this functionality can be implemented.

### Describe the solution you'd like

Introduce the following header actions in the header setter extension: 
- `insert`: Inserts the new header if it does not exist.
- `update`: Updates the header value if it exists.
- `upsert`: Inserts a header if it does not exist and updates the header if it exists.
- `delete`: Deletes the header.

Configuration example:

```
extensions:
  headers_setter:
    headers:
      - key: <header>
        action: insert
        from_context: <value from context>
      - key: <header>
        action: upsert
        value: <static value>
      - key: <header>
        action: update
        from_context: <value from context>
      - key: <header>
        action: delete    
```

### Describe alternatives you've considered

_No response_

### Additional context

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7596

### PR DESCRIPTION
**Pull request by kovrus**
**Description:** <Describe what has changed.>

Add header modification actions to the headers setter extension.

**Link to tracking Issue:** <Issue number if applicable>
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/16581
Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7596

**Testing:** <Describe what testing was performed and which tests were added.>
Unit tests

**Documentation:** <Describe the documentation added.>
Updated README

### CODE REVIEW

File: .chloggen/add-actions-to-headerssetter.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+note: Extend the headers setter extension with header modification actions.
+
+# One or more tracking issues related to the change
+issues: [16581, 7596]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  Update the existing headers setter configuration with `action: insert`.
```
Comment (atoulme) [inline .chloggen/add-actions-to-headerssetter.yaml:17]: Is this correct? I'd say to update, you'd use the action `update`, correct?
Do we want to mention the other actions too?
```suggestion
  Update the existing headers setter configuration:
        - `insert`: Inserts the new header if it does not exist.
        - `update`: Updates the header value if it exists.
        - `upsert`: Inserts a header if it does not exist and updates the header
          if it exists.
        - `delete`: Deletes the header.
```

**Review by atoulme — APPROVED** [review]
LGTM!

File: .chloggen/add-actions-to-headerssetter.yaml (hunk 1 above)
Comment (kovrus) [reply .chloggen/add-actions-to-headerssetter.yaml:17]: Right, it is `update`, I'll push a fix-up.

File: extension/headerssetterextension/config.go [hunk 2]
```diff
@@ -33,25 +34,51 @@ type Config struct {
...
+	DELETE actionValue = "delete"
+)
+
 // Validate checks if the extension configuration is valid
 func (cfg *Config) Validate() error {
 	if cfg.HeadersConfig == nil || len(cfg.HeadersConfig) == 0 {
 		return errMissingHeadersConfig
 	}
 	for _, header := range cfg.HeadersConfig {
+		if header.Action == "" {
```
Comment (jpkrohling) [inline extension/headerssetterextension/config.go:67]: This makes the extension backward incompatible. You can assume that all existing actions are "insert", and you'll be backward compatible.

File: extension/headerssetterextension/README.md [hunk 3]
```diff
@@ -17,27 +17,40 @@ header to the value extracted from the context.
...
-    - `key`: the header name
-    - `value`: the header value is looked up from the `value` property of the
-       extension configuration
-    - `from_context`: the header value is looked up from the request metadata,
-       such as HTTP headers, using the property value as the key (likely a header name)
+- `headers`: a list of header configuration objects that specify headers and
+  their value sources. Each configuration object has the following properties:
+    - `key`: The header name.
+    - `action`: An action to perform on the header. Supported actions are:
+        - `insert`: Inserts the new header if it does not exist.
```
Comment (jpkrohling) [inline extension/headerssetterextension/README.md:24]: If you decide to make this the default action, mark it as such here.

File: extension/headerssetterextension/config.go (hunk 2 above)
Comment (kovrus) [reply extension/headerssetterextension/config.go:67]: It is still `alpha`, so there was no intention to stay compatible. Therefore, I left a note in the change log about changing the existing config. Also, I wanted to require the action explicitly without taking `insert` or `upsert` as a default action.

File: extension/headerssetterextension/README.md (hunk 3 above)
Comment (kovrus) [reply extension/headerssetterextension/README.md:24]: I'd be inclined to always require an action actually.

File: extension/headerssetterextension/config.go (hunk 2 above)
Comment (kovrus) [reply extension/headerssetterextension/config.go:67]: maybe then `change_type: 'enhancement'` should be `breaking`?

**Comment by kovrus** [comment]
@jpkrohling can you take another look? I replied to some comments and updated the change log entry.

File: extension/headerssetterextension/config.go (hunk 2 above)
Comment (jpkrohling) [reply extension/headerssetterextension/config.go:64]: Can we do this in two steps then? The first is non-breaking, assuming that 'insert' is the default action, and will log a WARN if no action is defined, then on N+2, we fail if no action is provided. 

This way, current users won't be caught by surprise when they just bump the collector version: we warned them via changelog *and* via a WARN log entry on their collectors for at least one version.

**Review by jpkrohling — COMMENTED** [review]
LGTM, we need only to settle on the backward compatibility discussion.

**Comment by kovrus** [comment]
> Can we do this in two steps then? The first is non-breaking, assuming that 'insert' is the default action, and will log a WARN if no action is defined, then on N+2, we fail if no action is provided.

Made `upsert` the default action, so no breaking changes now. I'll follow up after a couple of releases with removing the default action.

File: extension/headerssetterextension/extension.go [hunk 4]
```diff
@@ -48,7 +50,22 @@ func newHeadersSetterExtension(cfg *Config) (auth.Client, error) {
...
+			a = action.Insert{Key: *header.Key}
+		case UPSERT:
+			a = action.Upsert{Key: *header.Key}
+		case UPDATE:
+			a = action.Update{Key: *header.Key}
+		case DELETE:
+			a = action.Delete{Key: *header.Key}
+		default:
+			a = action.Upsert{Key: *header.Key}
+			logger.Warn("Please set the header action explicitly")
```
Comment (jpkrohling) [inline extension/headerssetterextension/extension.go:66]: I would have added the default action directly to the `createDefaultConfig` and let the remaining of the code require upsert. It should be easier to change in the future.

Also, if we are planning on requiring the action to be explicitly set in a future version, we should already let users know right now. Like: `The action was not provided, using 'upsert'. In future versions, we'll require this to be explicitly set`

Comment (kovrus) [reply extension/headerssetterextension/extension.go:66]: > I would have added the default action directly to the createDefaultConfig and let the remaining of the code require upsert. It should be easier to change in the future.

The config is a list of `HeaderConfig`, so I am not sure whether we can set the default with `createDefaultConfig` then. We can set the default action in `Validate` but there is no access to the logger and validate mutates the config which is probably ok, but I preferred to put it in the `newHeadersSetterExtension` factory method.

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**